keps/sig-auth/20191101-add-request-id-to-k8s-logs.md

```diff
@@ -0,0 +1,330 @@
+creation-date: 2019-11-01
+last-updated: 2020-03-17
+status: provisional
+---
+
+# Add Request-ID to each Kubernetes component log
+
+## Table of Contents
+
+<!-- toc -->
+ - [Summary](#summary)
+ - [Motivation](#motivation)
+   - [Target User](#target-user)
+   - [Target User's objective](#target-users-objective)
+   - [Case 1](#case-2)
+   - [Case 2](#case-1)
+   - [Goals](#goals)
+   - [Non-Goals](#non-goals)
+ - [Proposal](#proposal)
+   - [Basic policy of Request-ID](#basic-policy-of-request-id)
+   - [Design Overview](#design-overview)
+     - [Design of Propagate Request-ID](#design-of-propagate-request-id)
+     - [Design of Export Request-ID](#design-of-export-request-id)
+     - [Design overview of Control Request-ID](#design-overview-of-control-request-id)
+     - [Detail design of Control Request-ID](#detail-design-of-control-request-id)
+ - [Test Plan](#test-plan)
+ - [Migration / Graduation Criteria](#migration--graduation-criteria)
+   - [Alpha](#alpha)
+   - [Beta](#beta)
+   - [GA](#ga)
+
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes a new unique logging meta-data into key Kubernetes logs(Details are described in [Design of Propagate Request-ID section](#design-of-propagate-request-id)). It makes us
+more easy to identify specific logs related to a single user operation (such as
+`kubectl apply -f <my-pod.yaml>`). This feature is similar to
+[Global request ID](https://docs.openstack.org/api-guide/compute/faults.html) for
+OpenStack. It greatly reduces investigation cost.
+
+## Motivation
+
+### Target User
+
+Support team in k8s Service Provider
+
+### Target User's objective
+
+We'd like to resolve quickly for end users' problem.
+
+Tracking logs among each Kubernetes component related to specific an user operation is very tough work. It is necessary to match logs by basically using timestamps and object's name as hints. If multiple users throw many API requests at the same time, it is very difficult to track logs across each Kubernetes component log. 
+
+It is difficult that support team in k8s Service Provider resolve end user's problem quickly in the above. Therefore, we'd like to add a new identifier which is unique to each user operation. This feature is useful for the following use cases:
+
+#### Case 1
+
+In case of insecure or unauthorized operation happens, it is necessary to
+identify what effect that operation caused. This proposed feature helps identify
+what happened at each component or server by each insecure / unauthorized API
+request. We can collect these logs as an evidence. This is similar to the 
+[Auditing](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/), 
+except for the following points.
+
+ - Audit only collects information about http request sending and receiving in kube-apiserver, so it can't track internal work of each component.
+ - Audit logs can't be associated to logs related to user operation (kubectl operation), because auditID is different for each http request.
```

Comment:
Please also note that context-based logging doesn't have nearly the same degree of security that audit logs have.  The context propagation described in the tracing proposal doesn't guarantee that all actions taken by controllers will be tied back to the same trace id.  In particular, if the trace context is overwritten, new spans/logs from components will be associated with the new trace context.  We can't provide guarantees around the completeness of logs for a given trace id.    Audit logging is also purposefully designed to be secure.  Component logs do not provide the same protections.

Reply:
> In particular, if the trace context is overwritten, new spans/logs from components will be associated with the new trace context.

Does this mean below? (I referred your KEP)

> High-level processes, such as starting a pod or restarting a failed container, could be interrupted before completion by an update to the desired state. While this leaves a "partial" trace for the first process, it is the most accurate representation of the work and timing of reconciling desired and actual state.

Even If logs before `overwriting trace-context` have same trace-ID and associated logs after overwriting have **another** trace-ID, I think both logs are valuable from a troubleshooting standpoint.